src/linker/Linker/MessageContainer.cs

```diff
@@ -46,8 +47,12 @@ namespace Mono.Linker
 		/// <returns>New MessageContainer of 'Error' category</returns>
 		public static MessageContainer CreateErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null)
 		{
-			if (!(code >= 1000 && code <= 2000))
+			if ((Assembly.GetExecutingAssembly () == Assembly.GetCallingAssembly ()) &&
```

Comment:
This will work only as long we keep the LinkContext.LogWarning and similar methods NOT used by custom steps. They're not in the ref assembly, but they're public. Maybe we should make them internal and explicitly comment that we don't want them to be made available to custom steps.

That said those methods are very helpful, so I'm hesitant to do that. Not sure about this yet.

Reply:
I guess if we expose those methods we could also hook them up into this logic - so it's not a blocker.
In that case I would just make them internal and add comment for now - noting that if we do expose them, we would need to correctly wire up the logic so that this keeps working as expected.

Reply:
I think this is over complicated. We should just expose the method differently by having 2 different methods.

```c#
public static MessageContainer CreateErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null); // 1000-2000 range

public static MessageContainer CreateCustomErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null) // exposed in ref with check for 6000+ range

```

Reply:
I like that - we could then also make the `CreateErrorMessage` internal (and exclude it from the ref assembly) and make the other one public and include it in ref. We could also add an assert with the code here to validate that the external one is not called from within the linker.

Reply:
Adding a new method for creating custom messages makes sense (this saves us from any perf impact), although then there's no point on keeping `CreateErrorMessage` in the ref assembly, we could just remove it.